### DISCUSSION TITLE
ast.Walk() and removing nodes

### DISCUSSION THREAD

**Discussion by zephryn** [yuin/goldmark#303] [Q&A]
I've been using `ast.Walk()` to comb through a parsed markdown document and extract (and in cases relevant to my issue, *remove*) nodes of certain types under certain conditions.

Here's a sample of the sort of thing I'm trying to do:
```go
doc := gm.Parser().Parse(text.NewReader(src))

var(
	foundImage bool
	foundHeading bool
)

ast.Walk(doc, func(node ast.Node, enter bool) (ast.WalkStatus, error) {
	switch child := node.(type) {
	case *ast.Heading:
		if child.Level == 1 && !foundHeading {
			child.Parent().RemoveChild(child.Parent(), child)
			foundHeading = true
		}
	case *ast.Image:
		if !foundImage {
			child.Parent().RemoveChild(child.Parent(), child)
			foundImage = true
		}
	}
	return ast.WalkContinue, nil
})
```
It might be important to note that I'm using a type switch over `child.Kind()` in this sample because of additional processing involved with specific node types in my actual code (e.g. `child.Destination` for images).

It seems like the walker coming across a heading before an image comes up results in only the heading being walked through, but the other way around results in correct output.

I kind of assume it's some anomaly I've managed to overlook with my half-awake programmer tunnel-vision, but I am interested if anyone has any advice, solutions or can otherwise comment on this before I get ahead of myself too much.

**Comment by reeseovine** [comment]
i was trying to do something similar. i'm rendering markdown for a git web frontend and i wanted to make some modifications. this code works for me:

```go
type rewriter struct {
	Repo string
	Ref  string
}
func (t *rewriter) Transform(doc *ast.Document, reader text.Reader, pctx parser.Context) {
	ast.Walk(doc, func(node ast.Node, enter bool) (ast.WalkStatus, error) {
		if !enter {
			return ast.WalkContinue, nil
		}

		kind := node.Kind().String()
		switch kind {
		case "Image":
			// Rewrite relative image source to point to raw file within the repo
			img, _ := node.(*ast.Image)
			dest := string(img.Destination)
			if !strings.HasPrefix(dest, "http") && !strings.HasPrefix(dest, "/") {
				img.Destination = []byte("/"+t.Repo+"/raw/"+t.Ref+"/"+dest)
			}
		case "TableHeader":
			// Remove empty table headers
			text := node.Text([]byte{})
			if len(text) == 0 {
				node.Parent().RemoveChild(node.Parent(), node)
			}
		}

		return ast.WalkContinue, nil
	})
}
```

i am loading the `rewriter` struct with the proper values before calling `Transform()`.

it also works regardless of the case statement order.

i'm over a year late but i hope this helps someone!

**edit:** `text := node.Text([]byte{})` doesn't appear to be the correct way to get a node's text value. it has no bearing on the relevance to the question, but i thought i should point it out. i'm open to suggestions for how to do it properly!

**Comment by j2kun** [comment]
This is because the walk function in goldmark is too simplistic:

```
for c := n.FirstChild(); c != nil; c = c.NextSibling() {
	if st, err := walkHelper(c, walker); err != nil || st == WalkStop {
		return WalkStop, err
	}
}
```

If you remove `c` from its parent, then `c.NextSibling` becomes nil and the walk terminates.

My workaround in https://github.com/gohugoio/hugo-goldmark-extensions/pull/6 is to anchor the walk on the parent of the node you want to remove, and walk the children like this:

```go
for c := n.FirstChild(); c != nil; {
	// Have to eagerly fetch this because `c` may be removed from the tree,
	// destroying its link to the next sibling.
	next := c.NextSibling()
	if c.Kind() == ast.KindParagraph {
		...
		n.RemoveChild(n, c)
	}
	c = next
}
```
